File: backend/app/api/schemas/lesson.py
from datetime import datetime
from typing import Literal
import uuid
from pydantic import BaseModel, Field


class LessonCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=255)
    subject: str | None = Field(None, max_length=100)
    language: Literal["ru", "kk", "en"]
    source: Literal["live", "upload"]
    group_id: uuid.UUID | None = None
    expected_terms: list[str] = Field(default_factory=list)
    roi: dict | None = None
    consent_confirmed: bool = Field(
        False,
        description="Must be true for live lessons according to Section 15.3 of TZ",
    )


class LessonResponse(BaseModel):
    id: uuid.UUID
    org_id: uuid.UUID
    group_id: uuid.UUID | None = None
    teacher_id: uuid.UUID
    title: str
    subject: str | None = None
    language: str
    source: str
    status: str
    expected_terms: list[str] = []
    roi: dict | None = None
    livekit_room: str | None = None
    last_seq: int
    cost_usd: float
    started_at: datetime | None = None
    ended_at: datetime | None = None
    created_at: datetime
    livekit_url: str | None = None
    teacher_token: str | None = None

    model_config = {"from_attributes": True}


class BlockUpdateRequest(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=150)
    body_md: str | None = Field(None, min_length=1)


class LessonStateResponse(BaseModel):
    lesson: LessonResponse
    last_seq: int
    blocks: list[dict] = []
    frames: list[dict] = []
    summary: dict | None = None
    transcript: list[dict] = []


class UploadUrlRequest(BaseModel):
    filename: str = Field(..., min_length=3, max_length=255)
    content_type: str = Field(default="video/mp4")


class UploadUrlResponse(BaseModel):
    upload_url: str
    s3_key: str
    expires_in: int


class ProcessLessonRequest(BaseModel):
    s3_key: str | None = None

File: frontend/src/types/lesson.ts
export type Language = 'ru' | 'kk' | 'en';
export type LessonSource = 'live' | 'upload';
export type VisibilityMode = 'live' | 'moderated';
export type LessonStatus = 'created' | 'live' | 'processing' | 'ready' | 'failed';
export type BlockStatus = 'approved' | 'pending_review';

export interface Annotation {
  id: string | number;
  label: string;
  type: 'box' | 'arrow' | 'highlight' | 'point';
  box_2d?: [number, number, number, number]; // [ymin, xmin, ymax, xmax] 0-1000
  point_2d?: [number, number]; // [y, x] 0-1000
  color?: string;
}

export interface Frame {
  id: string;
  lesson_id: string;
  t_ms: number;
  status: 'candidate' | 'selected' | 'rejected';
  s3_key: string;
  s3_key_annotated?: string;
  original_url?: string;
  annotated_url?: string;
  width: number;
  height: number;
  kind?: string;
  title?: string;
  ocr_markdown?: string;
  description?: string;
  annotations: Annotation[];
}

export interface Slide {
  id: string;
  lesson_id: string;
  slide_idx: number;
  s3_key: string;
  url?: string;
  phash: string;
  extracted_text: string;
  terms: string[];
  width: number;
  height: number;
}

export interface Callout {
  type: 'definition' | 'example' | 'warning' | 'important';
  title?: string;
  text: string;
}

export interface MediaArtifact {
  type: 'slide' | 'code' | 'diagram' | 'photo' | 'audio';
  title?: string;
  code_snippet?: string;
  language?: string;
  image_url?: string;
  audio_duration?: string;
  caption: string;
  badge?: string;
  align?: 'left' | 'right';
}

export interface NoteBlock {
  id: string;
  lesson_id: string;
  position: number;
  t_start_ms: number;
  t_end_ms: number;
  title: string;
  summary?: string;
  body_md: string;
  status?: BlockStatus;
  key_terms?: Array<{ term: string; definition: string }> | string[];
  callouts?: Callout[];
  frame_refs?: string[];
  uncertain?: string[];
  version: number;
  edited_by_teacher?: boolean;
  media_artifact?: MediaArtifact;
}

export interface LessonSummary {
  lesson_id: string;
  tldr?: string;
  outline?: Array<{ title: string; t_start_ms?: number }>;
  glossary?: Array<{ term: string; definition: string }>;
  takeaways?: string[];
  homework?: string;
}

export interface WordTimestamp {
  word: string;
  start_ms: number;
  end_ms: number;
}

export interface TranscriptSegment {
  id: string;
  lesson_id: string;
  start_ms: number;
  end_ms: number;
  text: string;
  speaker?: string;
  confidence?: number;
  words?: WordTimestamp[];
}

export interface Lesson {
  id: string;
  org_id: string;
  group_id?: string | null;
  teacher_id: string;
  title: string;
  subject?: string | null;
  language: Language;
  source: LessonSource;
  status: LessonStatus;
  visibility_mode?: VisibilityMode;
  expected_terms: string[];
  roi?: Record<string, any> | null;
  livekit_room?: string | null;
  last_seq: number;
  cost_usd: number;
  started_at?: string | null;
  ended_at?: string | null;
  created_at: string;
  livekit_url?: string | null;
  teacher_token?: string | null;
}

export interface LessonState {
  lesson: Lesson;
  last_seq: number;
  blocks: NoteBlock[];
  frames: Frame[];
  slides?: Slide[];
  summary?: LessonSummary | null;
  transcript: TranscriptSegment[];
}
